### PR TITLE
SD-560: implement basic PayoutManager contract

### PR DESCRIPTION
**Pull request by domagojlatecki**

### CODE REVIEW

File: test/services/payout-manager-test.ts [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+import { ethers } from "hardhat";
+import { BigNumber, ContractTransaction, Signer } from "ethers";
+import { expect } from "chai";
+import { describe, it } from "mocha";
+import { IERC20, MerkleTreePathValidator, PayoutManager } from "../../typechain";
+import * as helpers from "../../util/helpers";
+
+describe("Payout Manager test", function () {
+
+    //////// CONTRACTS ////////
```
Comment (githinho) [inline test/services/payout-manager-test.ts:11]: Remove this line, duplicated.

File: contracts/services/PayoutManager.sol [hunk 2]
```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./IMerkleTreePathValidator.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+
+contract PayoutManager {
+
+    using SafeMath for uint256;
```
Comment (githinho) [inline contracts/services/PayoutManager.sol:10]: You don't need `SafeMath` after solidity version 0.8.0.

File: contracts/services/PayoutManager.sol [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        require(_totalRewardAmount > 0, "PayoutManager: cannot create payout without reward");
+        require(_assetSnapshotMerkleDepth > 0, "PayoutManager: Merkle tree depth cannot be zero");
+
+        address payoutOwner = msg.sender;
+
+        // verify that payout owner approved enough reward asset
+        require(_totalRewardAmount <= _rewardAsset.allowance(payoutOwner, address(this)), "PayoutManager: insufficient reward asset allowance");
+
+        // transfer reward asset
+        _rewardAsset.transferFrom(payoutOwner, address(this), _totalRewardAmount);
```
Comment (githinho) [inline contracts/services/PayoutManager.sol:129]: Use transfer functions at the end. First, all require then state-changing functions and after all, comes transfer. Put this above emit event.

File: test/services/payout-manager-test.ts [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+        );
+
+        // verify funds are returned to owner
+        let ownerBalance = await rewardAsset.balanceOf(ownerAddress);
+        expect(ownerBalance).to.be.equal(oneToOneReward);
+
+        // verify remaining reward amount is now zero and payout is canceled
+        let afterPayoutInfo = await payoutManager.getPayoutInfo(payout.id);
+        verifyPayoutInfo(afterPayoutInfo, {...payout, canceled: true, remainingReward: 0});
+    });
```
Comment (githinho) [inline test/services/payout-manager-test.ts:444]: Maybe add the action that the holder tries to claim payout, this should fail.

Comment (githinho) [reply test/services/payout-manager-test.ts:444]: This is covered in the test below, nice!

File: test/services/payout-manager-test.ts [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+        verifyPayoutInfo(payoutInfo, payout);
+
+        // verify payout info per asset
+        let assetPayouts = await payoutManager.getPayoutsForAsset(asset1.address);
+        expect(assetPayouts.length).to.be.equal(1);
+        verifyPayoutInfo(assetPayouts[0], payout);
+
+        // verify that account not included in payout cannot claim reward
+        let nonIncludedAddress = await alice.getAddress();
+        let claim = payoutManager.connect(alice).claim(payout.id, nonIncludedAddress, 2500, proofs[0]);
```
Comment (githinho) [inline test/services/payout-manager-test.ts:754]: A lot `let` can be changed to `const`.

**Review by githinho — CHANGES_REQUESTED** [review]

File: contracts/services/PayoutManager.sol (hunk 2 above)
Comment (domagojlatecki) [reply contracts/services/PayoutManager.sol:10]: Good to know 👍

**Review by githinho — APPROVED** [review]

File: contracts/services/PayoutManager.sol [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        // validate Merkle proof to check if payout should be made for (address, balance) pair
+        bool containsNode = merkleTreePathValidator.containsNode(
+            payout.assetSnapshotMerkleRoot,
+            payout.assetSnapshotMerkleDepth,
+            _wallet,
+            _balance,
+            _proof
+        );
+
+        require(containsNode, "PayoutManager: requested (address, blaance) pair is not contained in specified payout");
```
Comment (filipdujmusic) [inline contracts/services/PayoutManager.sol:191]: typo blaance

**Review by filipdujmusic — CHANGES_REQUESTED** [review]
Overall very good job ! ⭐️
The contracts are well structured , the code looks nice and the tests are very well organized covering all the important scenarios.

File: contracts/services/PayoutManager.sol [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    mapping(uint256 => Payout) private payoutsById;
+    mapping(address => uint256[]) private payoutsByAssetAddress;
+    mapping(uint256 => mapping(address => bool)) private payoutClaims;
+
+    //------------------------
+    //  EVENTS
+    //------------------------
+    event PayoutCreated(uint256 payoutId, address indexed payoutOwner, IERC20 asset, IERC20 rewardAsset, uint256 totalRewardAmount);
+    event PayoutCanceled(uint256 payoutId, IERC20 asset);
+    event PayoutClaimed(uint256 payoutId, address indexed wallet, uint256 balance, uint256 payoutAmount);
```
[the remark this answers is not in the document]
Comment (githinho) [reply contracts/services/PayoutManager.sol:47]: Looks good to me.

**Comment by domagojlatecki** [comment]
This pull request is temporarily on hold as there need to be some changes related to payouts as discussed with @mightymatth and @filipdujmusic. Another pull request will be made which will merge into this one to implement new changes.

**Pull request merged**